### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by A-Tanner**
Extra Credit Attempted. JSON is handled using JsonJokeService, which is derived from JokeService.

### CODE REVIEW

**Comment by GavenElder** [comment]
I will PR this.

**Comment by GavenElder** [comment]
For this assignment do the following.
    Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔

    The JokeService will need to have the interface applied to it. ✔
    Create an implementation of the output interface that writes out the joke to the console. ✔
    Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
    The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔

    Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
    Enable nullability in the .csproj files and ensure no errors or warnings: ✔

Fundamentals

    Ensure you enable:
        nullable reference types is enabled ✔
        net6 targetted ✔
        C# 10.0 ✔
        and enabled .NET analyzers for both projects ✔

    For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
    (I didn't see the use Assert.AreEqual<T>() within assignment.)

    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔
   (besides for the exception of the "coverage.cobertura.xml" file) This one is rather lengthy and complicated.)

Extra Credit

    Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)?
    The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string.  ❌

Doesn't seem like the extra credit portion is completely done. But has been attempted.

**Comment by A-Tanner** [comment]
Hey Gavin! Thanks for the peer review.

I had a couple of questions ;

What exception are you getting for coverage.cobertura.xml? This is intended to be used with coverlet and should not cause any program level exceptions.

Which portion of the extra credit was uncompleted? As far as I can see, The capacity of the ConsoleWriter to write to console has been tested and the JsonJokeService is able to retrieve and translate jokes from the JSON encoding of the API.

Finally, note that the way you conducted the peer review has marked you as a contributor, and not as a reviewer

**Comment by GavenElder** [comment]
Hello!
What I meant by "exception" is that it looks rather complicated, your code works great with it! I probably should have clarified or used a different word rather than "exception": to describe that! 

I went back and ran your extra credit portion and you are right, it does run as intended. It must; have been something on my end!

Cheers!

**Comment by A-Tanner** [comment]
That all makes sense, no worries. Thanks for getting back to me about the specifics and even re-running my code!

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 1]
```diff
@@ -2,14 +2,22 @@
...
-        private HttpClient HttpClient { get; } = new();
+
+        public string? URL { get; private set; }
+        protected HttpClient HttpClient { get; } = new();
+
 
         public string GetJoke()
         {
-            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api").Result;
+            string joke = HttpClient.GetStringAsync(URL).Result;
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JokeService.cs:14]: Although `GetStringAsync` allows for a null string, you may still want to do a null check here. I'd imagine you would expect the user to call `SetupService` before calling this method?

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj [hunk 2]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/CanHazFunny.csproj:3]:
```c#
        static void Main()
        {
            //Feel free to use your own setup here - this is just provided as an example
            var jokeService = new JokeService();
            jokeService.SetupService("https://geek-jokes.sameerkumar.website/api");
            new Jester(jokeService, new ConsoleOutput()).TellJoke();
        }
```

You should configure your `Main` method in `Program.cs` to do something when the program starts up. For example, the above would be acceptable. Or you can call the json service.

File: CanHazFunny/CanHazFunny/JsonJokeService.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            _Joke? joke = JsonSerializer.Deserialize<_Joke?>(HttpClient.GetStringAsync(URL).Result);
+            return joke?.joke;
+        }
+
+
+        private class _Joke
+        {
+            public string? joke { get; set; }
+        }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JsonJokeService.cs:24]: I would move this class to be outside of the JsonJokeService class. Aside from that, `joke` should be pascal case. The issue of course, is it doesn't deserialize properly that way.

To fix that you could import `Newtonsoft.Json` and do the following...
```suggestion
        private class _Joke
        {
            [JsonProperty("joke")]
            public string? Joke { get; set; }
        }
```

**Pull request closed without merging**